### ISSUE TITLE
test failure in 2.4.2

### ISSUE THREAD

**Issue by doko42** [ronf/asyncssh#325]
trying to build and run the tests, I see one test failure with this release, both for Python 3.8.6 and 3.9.0:

======================================================================
ERROR: test_confirm (tests.test_agent._TestAgent)
Test confirmation of key
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/tests/util.py", line 69, in async_wrapper
    return self.loop.run_until_complete(coro(self, *args, **kwargs))
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/<<PKGBUILDDIR>>/tests/test_agent.py", line 52, in agent_wrapper
    await func(self, agent)
  File "/<<PKGBUILDDIR>>/tests/test_agent.py", line 379, in test_confirm
    sig = await agent_key.sign(b'test')
  File "/<<PKGBUILDDIR>>/asyncssh/agent.py", line 147, in sign
    return await self._agent.sign(self.public_data, data, self._flags)
  File "/<<PKGBUILDDIR>>/asyncssh/agent.py", line 271, in sign
    raise ValueError('Unable to sign with requested key')
ValueError: Unable to sign with requested key

----------------------------------------------------------------------
Ran 1198 tests in 612.215s

FAILED (errors=1, skipped=65)

complete build log at
https://launchpad.net/ubuntu/+source/python-asyncssh/2.4.2-0ubuntu3

**Comment by ronf** [comment]
Thanks for the report. Unfortunately, I'm not seeing that problem here, even on Ubuntu. I'm guessing there may be an environmental difference related to the way my test code is attempting to mock out the SSH "askpass" functionality, but without more info that'll be hard to track down.

Do you have an environment you can use to reproduce this that I could either remotely access or download as something like a Docker container?

The error "Unable to sign with requested key" doesn't tell us much here, unfortunately. That just means that the OpenSSH ssh-agent that was being tested against returned SSH_AGENT_FAILURE on a call to sign using one of its keys, but the ssh-agent protocol doesn't really provide any detail about the reason for a failure.

**Comment by doko42** [comment]
https://launchpad.net/~doko/+archive/ubuntu/ppa/+packages

indeed, that succeeds on 20.04 and 20.10, but not in the current development release (21.10).
cd images are avalable in https://cdimage.ubuntu.com/daily-live/current/
but maybe you see something obvious in the build dependencies in the above logs?

**Comment by ronf** [comment]
Thanks. The good news is that I've been able to reproduce the problem using a 21.04 (Hirsute) live CD from the link above.

So far, I haven't been able to determine why it is failing, but it definitely seems related to not launching the "ssh-askpass" script that is used in the unit testing to simulate a user confirming whether or not to allow signing with a particular key. For some reason, it seems like ssh-agent is not able to run that script properly when attempting to perform the confirmation.

**Comment by ronf** [comment]
Ok - I think I found the issue. It is related to a change made in ssh-agent in OpenSSH 8.4p1. Previously, when ssh-agent looked to see if the DISPLAY environment variable was set in read_passphrase(), it allowed an empty string as the value. However, as part of some changes to that function in 8.4, it now checks not only that DISPLAY is present in the environment but also that it is a non-empty string.

In my unit test code, I set DISPLAY in the environment to make sure that it would try to run my stub SSH_ASKPASS binary. However, I set it to the empty string. This worked in early OpenSSH releases, but now an empty string for DISPLAY prevents SSH_ASKPASS from being run in 8.4p1.

The good news is that this is easy to work around, but I wonder if there are other people out there relying on some kind of scripted version of ssh-askpass that doesn't involve real user interaction which might get burned by this.

**Comment by ronf** [comment]
Ok - a fix for this is available as commit 8581f5c in the "develop" branch.

If you want to patch it yourself, you can apply the following:

```patch
diff --git a/tests/test_agent.py b/tests/test_agent.py
index ea6f68c..4e98fbd 100644
--- a/tests/test_agent.py
+++ b/tests/test_agent.py
@@ -111,7 +111,7 @@ class _TestAgent(AsyncTestCase):
     async def asyncSetUpClass(cls):
         """Set up keys and an SSH server for the tests to use"""
 
-        os.environ['DISPLAY'] = ''
+        os.environ['DISPLAY'] = ' '
         os.environ['HOME'] = '.'
         os.environ['SSH_ASKPASS'] = os.path.join(os.getcwd(), 'ssh-askpass')
 
```

**Comment by doko42** [comment]
thanks! that fixes the issue.

**Comment by ronf** [comment]
This fix is now part of AsyncSSH 2.5.0.
